### PR TITLE
Adds TaskScheduler base class for sync/async task scheduling.

### PR DESCRIPTION
**Pull request by andescu**
To be able to remove the std::thread().detach() usage this introduces
a base class named TaskScheduler for scheduling sync/async tasks, depending
on the user's implementation. As this is interface users are free to
implement their own task scheduling algorithm by overloading this class.

Relates-to: OLPEDGE-393

Signed-off-by: Andrei Popescu <andrei.popescu@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+ protected:
+  /**
+   * @brief Abstract enqueue task interface to be implemented by subclass.
+   *
+   * Implement this method in your subclass taking TaskScheduler as base and
+   * provide a custom algorithm for scheduling tasks enqueued by the SDK.
+   *
+   * @param[in] func Rvalue reference of the to be enqueued task. Move this task
```
Comment (serhiikobalt) [inline olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h]: typo here

File: olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ * that takes any callable target (lambda expression, bind expression or any
+ * other function object) as input and adds it to the execution pipeline.
+ *
+ */
+class CORE_API TaskScheduler {
+ public:
+  /// Alias for abstract interface input
+  using CallFuncType = std::function<void()>;
+
+  virtual ~TaskScheduler() = default;
```
Comment (serhiikobalt) [inline olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h]: Put destructor definition into separate .cpp . This way you can also mention it in CMakeLists and make sure it compiles

Comment (andescu) [reply olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h]: As this will be used pretty soon by the actual impl I assume this check will be soon performed. No need to place extra cpp just for a default destructor.

Comment (serhiikobalt) [reply olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h]:
```
[  4%] Building CXX object olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/src/thread/TaskScheduler.cpp.o
In file included from /mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/src/thread/TaskScheduler.cpp:1:
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h:62:12: error: 'this' argument to member function 'IsCancelled' has type 'const client::CancellationContext', but function is not marked const
      if (!context.IsCancelled()) {
           ^~~~~~~
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h:74:15: note: 'IsCancelled' declared here
  inline bool IsCancelled() {
              ^
In file included from /mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/src/thread/TaskScheduler.cpp:1:
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h:61:24: error: call to implicitly-deleted copy constructor of 'client::CancellationContext'
    auto task = [func, context]() {
                       ^~~~~~~
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h:80:24: note: copy constructor of 'CancellationContext' is implicitly deleted because field 'mutex_' has a deleted copy constructor
  std::recursive_mutex mutex_;
                       ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/mutex:101:5: note: 'recursive_mutex' has been explicitly marked deleted here
    recursive_mutex(const recursive_mutex&) = delete;
    ^
In file included from /mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/src/thread/TaskScheduler.cpp:1:
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h:67:12: error: call to implicitly-deleted copy constructor of 'client::CancellationContext'
    return context;
           ^~~~~~~
/mnt/c/here-olp-edge-sdk-cpp/olp-cpp-sdk-core/include/olp/core/client/CancellationContext.h:80:24: note: copy constructor of 'CancellationContext' is implicitly deleted because field 'mutex_' has a deleted copy constructor
  std::recursive_mutex mutex_;
                       ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/mutex:101:5: note: 'recursive_mutex' has been explicitly marked deleted here
    recursive_mutex(const recursive_mutex&) = delete;
    ^
3 errors generated.
olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/build.make:974: recipe for target 'olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/src/thread/TaskScheduler.cpp.o' failed
make[3]: *** [olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/src/thread/TaskScheduler.cpp.o] Error 1
CMakeFiles/Makefile2:503: recipe for target 'olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/all' failed
make[2]: *** [olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/all] Error 2
CMakeFiles/Makefile2:515: recipe for target 'olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/rule' failed
make[1]: *** [olp-cpp-sdk-core/CMakeFiles/olp-cpp-sdk-core.dir/rule] Error 2
Makefile:255: recipe for target 'olp-cpp-sdk-core' failed
make: *** [olp-cpp-sdk-core] Error 2
```

Comment (andescu) [reply olp-cpp-sdk-core/include/olp/core/thread/TaskScheduler.h]: Yes, as this https://github.com/heremaps/here-olp-edge-sdk-cpp/pull/37 is missing.

**Comment by andescu** [comment]
Rebased to include #37 
@serhiikobalt Please have another look.
